etl/etl/competition.py

```diff
@@ -0,0 +1,802 @@
+        to the competition"""
+        return []
+
+    def cell(self, proposal, column_name):
+        """Returns the new cell for the proposal PROPOSAL denoted by
+        COLUMN_NAME.  This is information generated by the Adder, but
+        can use other information within the PROPOSAL."""
+        pass
+
+
+class MediaWikiTitleAdder(InformationAdder):
+    """An InformationAdder that adds the MediaWiki Title column, which
+    is a wiki safe version of "<Project Title> (<Application #>)"""
+
+    title_column_name = "MediaWiki Title"
+
+    def __init__(self, project_column_name):
+        self.project_column_name = project_column_name
+
+    def column_type(self, column_name):
+        return None
+
+    def column_names(self):
+        return [self.title_column_name]
+
+    def cell(self, proposal, column_name):
+        import unidecode
+
+        title = "%s (%s)" % (proposal.cell(self.project_column_name), proposal.key())
+
+        # Project titles sometimes have characters that lead to malformed
+        # mediawiki titles so this code cleans that up.
+        for c in [
+            "#",
+            "<",
+            ">",
+            "[",
+            "]",
+            "{",
+            "|",
+            "}",
+        ]:
+            if title.find(c) != -1:
+                title = title.replace(c, "-")
+        while len(bytes(title, "UTF-8")) > 255:
+            title = title[:-1]
+        title = title.replace("\n", " ")
+
+        # Also convert non unicode because we do this with titles on the other side
+        return unidecode.unidecode_expect_nonascii(title).strip()
+
+
+class Attachment:
+    """Represents an attachment on the file system, and handles special cases
+    about display names to make sure things show up correctly"""
+
+    def __init__(self, key, filename, rank, column_name, path):
+        self.file = filename
+        self.rank = rank
+        self.key = key
+        self.path = path
+        self.column_name = column_name
+        self.name = re.sub("^\\d*_", "", filename)
+        self.name = re.sub("\.pdf$", "", self.name)
+        if len(self.name) > 33:
+            self.name = self.name[0:15] + "..." + self.name[(len(self.name) - 15) :]
+
+
+class BasicAttachments(InformationAdder):
+    """Represents a set of attachments generated from an ATTACHMENT_DIR provided
+    at initialization.  These attachments can later be pulled out to upload to
+    MediaWiki, but also is an InformationAdder so that those attachments values
+    are in the Proposal so that torque knows where they are.
+
+    The KEYS passed in the constructor refers to which proposals to look for
+    attachments in, as the ATTACHMENTS_DIR should always have subdirectories
+    per proposal by key.
+
+    The column names can be variable based on what attachments are important
+    for the competition.
+
+    The rank for all attachments is set to 99 for default.  This is used for
```

Comment:
Actually, this code breaks in a subtle way because of how the sort is done.  My kingdom for sort taking a comparison method.  I'll have to fix this more completely.

Reply:
Wait, Python sort does take a comparison method...  At least, I know I've done that in Python before.  Isn't it just the optional `key` parameter on the sorted() function and on sort() methods?